src/test/mir-opt/early_otherwise_branch.opt5.EarlyOtherwiseBranch.diff

```diff
@@ -0,0 +1,73 @@
+      let mut _0: u32;                     // return place in scope 0 at $DIR/early_otherwise_branch.rs:49:45: 49:48
+      let mut _3: (std::option::Option<u32>, std::option::Option<bool>); // in scope 0 at $DIR/early_otherwise_branch.rs:50:11: 50:17
+      let mut _4: std::option::Option<u32>; // in scope 0 at $DIR/early_otherwise_branch.rs:50:12: 50:13
+      let mut _5: std::option::Option<bool>; // in scope 0 at $DIR/early_otherwise_branch.rs:50:15: 50:16
+      let mut _6: isize;                   // in scope 0 at $DIR/early_otherwise_branch.rs:51:19: 51:26
+      let mut _7: isize;                   // in scope 0 at $DIR/early_otherwise_branch.rs:51:10: 51:17
+      let _8: u32;                         // in scope 0 at $DIR/early_otherwise_branch.rs:51:15: 51:16
+      let _9: bool;                        // in scope 0 at $DIR/early_otherwise_branch.rs:51:24: 51:25
++     let mut _10: isize;                  // in scope 0 at $DIR/early_otherwise_branch.rs:51:19: 51:26
++     let mut _11: bool;                   // in scope 0 at $DIR/early_otherwise_branch.rs:51:19: 51:26
+      scope 1 {
+          debug a => _8;                   // in scope 1 at $DIR/early_otherwise_branch.rs:51:15: 51:16
+          debug b => _9;                   // in scope 1 at $DIR/early_otherwise_branch.rs:51:24: 51:25
+      }
+  
+      bb0: {
+          StorageLive(_3);                 // scope 0 at $DIR/early_otherwise_branch.rs:50:11: 50:17
+          StorageLive(_4);                 // scope 0 at $DIR/early_otherwise_branch.rs:50:12: 50:13
+          _4 = _1;                         // scope 0 at $DIR/early_otherwise_branch.rs:50:12: 50:13
+          StorageLive(_5);                 // scope 0 at $DIR/early_otherwise_branch.rs:50:15: 50:16
+          _5 = _2;                         // scope 0 at $DIR/early_otherwise_branch.rs:50:15: 50:16
+          (_3.0: std::option::Option<u32>) = move _4; // scope 0 at $DIR/early_otherwise_branch.rs:50:11: 50:17
+          (_3.1: std::option::Option<bool>) = move _5; // scope 0 at $DIR/early_otherwise_branch.rs:50:11: 50:17
+          StorageDead(_5);                 // scope 0 at $DIR/early_otherwise_branch.rs:50:16: 50:17
+          StorageDead(_4);                 // scope 0 at $DIR/early_otherwise_branch.rs:50:16: 50:17
+          _7 = discriminant((_3.0: std::option::Option<u32>)); // scope 0 at $DIR/early_otherwise_branch.rs:51:10: 51:17
+-         switchInt(move _7) -> [1_isize: bb2, otherwise: bb1]; // scope 0 at $DIR/early_otherwise_branch.rs:51:10: 51:17
++         StorageLive(_10);                // scope 0 at $DIR/early_otherwise_branch.rs:51:10: 51:17
++         _10 = discriminant((_3.1: std::option::Option<bool>)); // scope 0 at $DIR/early_otherwise_branch.rs:51:10: 51:17
++         StorageLive(_11);                // scope 0 at $DIR/early_otherwise_branch.rs:51:10: 51:17
++         _11 = Ne(_10, _7);               // scope 0 at $DIR/early_otherwise_branch.rs:51:10: 51:17
```

Comment:
So... isn't this already a miscompilation? `bb2` is `(Err(_), Some(_)) => 0,` and `bb1` is `_ => 1`, but with the new logic here, we'd also go into `=> 0` for `(Ok(_), None)`.

Reply:
I think this is preexisting, the optimization also did this before this PR

Reply:
Ah no, it's not preexisting. The commit removing the additional branch causes this